### PR TITLE
#168781629 - Users Filter Requests

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Users Filter Requests.

#### Description of Task to be completed?
- Create searchBar Component with fields
- Create actions to fetch filtered requests

#### How should this be manually tested?
- Clone ft-filter-requests-168781629 branch
- run `npm i`
- run `npm run dev`
- Click on `Requests` in the MenuBar
- Click on `Filter Requests` to show/hide the Search Bar

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781629](https://www.pivotaltracker.com/story/show/168781629)

#### Screenshots (if appropriate)
Screen Shot 2019-12-19 at 18 03 09
Screen Shot 2019-12-19 at 18 03 22
Screen Shot 2019-12-19 at 18 09 44

#### Questions:
N/A

### CODE REVIEW

**Review by Cheza-Dzabala — COMMENTED** [review]
@nignanthomas, looks good.. However, a user should not be able to enter a request id from the front end

File: src/components/pages/requests/SearchComponent.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+import PropTypes from 'prop-types';
+import {
+  Container, Row, Col, Form, Button, Alert,
+} from 'react-bootstrap';
+import { connect } from 'react-redux';
+import { getLocations } from '../../../actions/locationActions';
+import { searchRequestAction, getRequestsAction } from '../../../actions/requestsActions';
+import Breadcrumbs from '../../global/Breadcrumbs';
+// import RequestItem from './RequestItem';
+
```
Comment (alainmateso) [inline src/components/pages/requests/SearchComponent.js]: remove commented code

File: src/components/pages/requests/SearchComponent.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    Object.keys(options).forEach((key) => {
+      if (options[`${key}`] !== '') {
+        searchParams += `${key}=${options[`${key}`]}&`;
+      }
+    });
+    searchParams = searchParams.substring(0, searchParams.length - 1);
+    if (searchParams !== '') {
+      searchParams = `?${searchParams}`;
+    }
+    console.log('searchParams ===> ', searchParams);
```
Comment (alainmateso) [inline src/components/pages/requests/SearchComponent.js]: remove console.log statements

File: src/components/pages/requests/SearchComponent.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            <Col xs={12} sm={6} md={4} lg={4}>
+              <Form.Group>
+                <Form.Control type="number" onChange={this.handleChange} name="duration" min="1" max="999" data-test="duration" placeholder="Duration in days..." />
+              </Form.Group>
+            </Col>
+
+            <Col xs={12} sm={6} md={4} lg={4}>
+              <Form.Group>
+                <Form.Control as="input" type="date" name="departureDate" onChange={this.handleChange} />
+                {/* <DayPickerInput /> */}
```
Comment (alainmateso) [inline src/components/pages/requests/SearchComponent.js]: remove commented code

File: src/views/requests/ViewRequests.js [hunk 4]
```diff
@@ -25,17 +26,49 @@ export class ViewRequests extends Component {
...
   render() {
     let data;
     let dataError;
     const { isLoading } = this.state;
     const { props } = this;
+    const { searchRequests } = props;
+    const { status, searchData, searchDataError } = searchRequests;
+    console.log('searchStatus ===> ', status);
+    console.log('searchData ===> ', searchData);
+    console.log('searchDataError ===> ', searchDataError);
```
Comment (alainmateso) [inline src/views/requests/ViewRequests.js]: remove console.log statements

File: src/views/requests/ViewRequests.js [hunk 5]
```diff
@@ -25,17 +26,49 @@ export class ViewRequests extends Component {
...
+
+    console.log('searchRequests ===> ', searchRequests);
```
Comment (alainmateso) [inline src/views/requests/ViewRequests.js]: remove console.log statements

File: src/views/requests/ViewRequests.js [hunk 6]
```diff
@@ -25,17 +26,49 @@ export class ViewRequests extends Component {
...
+    // if (props.searchData) {
+    //   searchData = props.searchData.data;
+    // }
+    // if (props.searchDataError) {
+    //   searchDataError = props.searchDataError;
+    // }
+    // if (status !== '') {
+    //   data = props.searchData;
+    //   dataError = props.searchDataError;
+    // }
```
Comment (alainmateso) [inline src/views/requests/ViewRequests.js]: remove commented code

File: src/views/requests/ViewRequests.js [hunk 7]
```diff
@@ -57,19 +90,28 @@ export class ViewRequests extends Component {
...
+        { status === '' && this.renderResults(data, dataError)}
+        { status !== '' && this.renderResults(searchData, searchDataError)}
+        {/* {data && data.length === 0 && <Alert variant="danger" heading="Error" message="No Requests Found" />}
         {data && data.map((item) => (
           <Row key={item.id}>
             <RequestItem item={item} />
           </Row>
         ))}
         <Row>
           {dataError && <Alert variant="danger" heading="Error" message={dataError.message} />}
```
Comment (alainmateso) [inline src/views/requests/ViewRequests.js]: remove commented code

File: src/views/requests/ViewRequests.js [hunk 8]
```diff
@@ -78,6 +120,13 @@ export class ViewRequests extends Component {
 export const mapStateToProps = (state) => ({
   data: state.requests.data,
   dataError: state.requests.dataError,
+
+  searchRequests: state.searchRequests,
+
+  // searchData: state.searchRequests.data,
+  // searchDataError: state.searchRequests.dataError,
+  // searchStatus: state.searchRequests.status,
+
```
Comment (alainmateso) [inline src/views/requests/ViewRequests.js]: remove commented code

**Review by alainmateso — CHANGES_REQUESTED** [review]
Great job @nignanthomas . Kindly work on the following feedback
- Since the id of the request is not displayed anywhere on the request information, kindly add it so that the search by id makes sense. 
- When no results are found, I believe it is better to use any other selector like just a label to tell the user that no requests were found instead of bring the alert component

I've also left a few comments, please check them out. Impressive work 👍🏾

**Review by alainmateso — APPROVED** [review]

File: src/components/pages/requests/SearchBar.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+  hideFilter = () => {
+    const { state: { openSearch } } = this;
+    this.setState((state) => ({ ...state, openSearch: !openSearch }));
+  };
+
+  searchRequest = async () => {
+    const { state, props } = this;
+    let searchParams = '';
+    const { isLoading, openSearch, ...options } = state;
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js:54]: Add a new line

File: src/components/pages/requests/SearchBar.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    Object.keys(options).forEach((key) => {
+      if (options[`${key}`] !== '') {
+        searchParams += `${key}=${options[`${key}`]}&`;
+      }
+    });
+    searchParams = searchParams.substring(0, searchParams.length - 1);
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js]: Add a new line

File: src/components/pages/requests/SearchBar.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+  searchRequest = async () => {
+    const { state, props } = this;
+    let searchParams = '';
+    const { isLoading, openSearch, ...options } = state;
+    Object.keys(options).forEach((key) => {
+      if (options[`${key}`] !== '') {
+        searchParams += `${key}=${options[`${key}`]}&`;
+      }
+    });
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js:60]: Add a new line

File: src/components/pages/requests/SearchBar.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    searchParams = searchParams.substring(0, searchParams.length - 1);
+    if (searchParams !== '') {
+      searchParams = `?${searchParams}`;
+    }
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js:65]: Add a new line

File: src/components/pages/requests/SearchBar.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    this.setState((state) => ({ ...state, isLoading: true }));
+    await props.searchRequestAction(searchParams);
+    this.setState((state) => ({ ...state, isLoading: false }));
+  }
+
+  render() {
+    const { props, state } = this;
+    const { isLoading, openSearch } = state;
+    const { locations } = props;
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js:75]: Add a new line

File: src/components/pages/requests/SearchBar.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        <Col xs={12} sm={6} md={4} lg={4}>
+          <Button className="full-width-buttons" style={{ width: '100%' }} onClick={() => this.hideFilter()} data-test="filter-request">
+            Filter Requests
+          </Button>
+        </Col>
+        </Container>
+      { openSearch && (
+        <Container className="search-bar">
+          <Row className="center-items">
+            {/* <Col xs={12} sm={6} md={4} lg={4}>
```
Comment (nakiwuge) [inline src/components/pages/requests/SearchBar.js]: Remove commented code

File: src/styles/requests/requests.scss [hunk 15]
```diff
@@ -1,5 +1,5 @@
 .request-item {
-  background-color: $light;
+  background-color: #ffffff;
```
Comment (nakiwuge) [inline src/styles/requests/requests.scss]: move this color to the variables

File: src/styles/breadcrumbs.scss [hunk 16]
```diff
@@ -1,3 +1,6 @@
 .breadcrumbs {
 	width: 40%;
 }
+.bc-bg {
+	background-color: #ffffff !important;
```
Comment (nakiwuge) [inline src/styles/breadcrumbs.scss]: move this color to the variables

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@nignanthomas this is great work however you have a date picker as one of the options but it doesn't have a label so a user cannot know whether its a date for creating a request or arrival date or destination date.

I have left other comments on the pr kindly check them out

**Review by Cheza-Dzabala — APPROVED** [review]

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**